Randomizer Auto Location tracker

sonoftunk opened the pull request:
Adds an Automatic Check Tracker, an enhancement of the Item Tracker. 
Props to PR #1117 for the idea and the inspiration!

- Currently it has all locations of Vanilla Dungeons, Overworld, Gold Skulltulas, Keys, Shops, etc. Based on the already implemented Location Exclusion list in Randomizer. 
- Checks are shown based on save file. As such, this is backwards compatible with any save file, as long as the save bits haven't changed between versions. Works best when Auto-Save is on. Uses 3drando as a basis for mapping checks to their save bits.
- Upon entering a new scene, the tracker will scroll to the area that contains that scene. 
- A search for unchecked locations
- There's also an `*` indicator for the last item retrieved.

![image](https://user-images.githubusercontent.com/59073987/200716778-891aef52-ce49-4656-b6a5-0d3736fcf23e.png)

Known issues:
- [x] Some Checks will not track. Still needs every check confirmed to work. 
- [x] ~In some settings, checks will be spoiled even without checking~ These should all be caught at this point, but not guaranteed
- [x] Some Checks are always visible, even if they can't be retrieved due to settings, or always invisible even when settings allow for it. For example, shopsanity always shown. Some are partially fixed, like Gerudo guards, but those still currently require a reboot of SoH before they show/hide properly. 
- [x] Language hardcoded English
- [x] Sometimes the `*` is flaky, disappearing and reappearing or showing on the wrong item. Always shows the wrong area on loading the save file
- [x] ~Only supports Randomizer, not vanilla OoT~ Won't do
- [x] ~Items that are switched with blue rupees show the item it would have been~ Won't do
- [x] ~Some checks save to the file prior to the item showing on screen, allowing a few seconds of spoiler when autosave enabled~ Won't Do

Future ideas:
- Support MQ dungeons
- Hint tracking. Showing hints as a part of the list of checks or beside the item they hinted.
- More States for checks. Hinted, Seen, Scummed, Checked. For example, once you've seen a free-standing item, it could show in a different colour/italics on the left side.
- Single column tracking mode. Order checks based on above states in each area. Give options for which to show and hide. 
- Log Mode. Shows entries as [Time] [Area] [Check] [Item]. Could easily see it took 7 minutes since you last got that Skulltula from the Song of Storms grotto chest in Kokiri forest. 

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/470953226.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/470953227.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/470953228.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/470953229.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/470953230.zip)
<!--- section:artifacts:end -->

sonoftunk committed:
Initial Auto-Location Tracking

sonoftunk committed:
Updates styling for checked items in Check Tracker

sonoftunk committed:
Fixes crashing issues with Check Tracker rendering

sonoftunk committed:
Auto scroll Check Tracker windows to current area

sonoftunk committed:
Checks Tracker only shows checked locations once in game

sonoftunk committed:
Fixes issue where Check Tracker would auto-scroll to wrong location when entering a grotto

sonoftunk committed:
Fixes gerudo fortress checks showing in tracker when settings have them removed

sonoftunk committed:
Merge remote-tracking branch 'origin/develop' into LocationTracker

sonoftunk committed:
InfTable Check Location Fixes: Lost Dog Richard, Rolling Gorons, HBA 1000, Thawed Zora

sonoftunk committed:
Fixes most Event Chk Inf Checks on Check Tracker, and Frogs

sonoftunk committed:
Check Tracker - Song from Malon now works

sonoftunk committed:
Check Tracker
- Fixes Claim Check check not tracking
- Fixes spoiler for Claim Check check when getting Biggoron's Sword
- Fixes LACS being spoiled and not tracking

sonoftunk committed:
Merge remote-tracking branch 'origin/develop' into LocationTracker

PurpleHato commented:
Wouldn't it be better to be it's own thing under the Randomizer Menu?
It doesn't make sense to me to have the Check Tracker being hidden inside the Item Tracker since they are two different things

With this current implementation, the user need to bring the item tracker settings, to then click a checkbox to show the Check Tracker If someone want to bring it up

It would be way easier to simply make it a button like the Item Tracker and Item Tracker Settings

sonoftunk committed:
Check Tracker - Fixes #2 for Skull Mask and Mask of Truth checks

sonoftunk committed:
Check Tracker
- Adds support for merchants
- Fixes #7 for Bomchu Salesman
- Fixes #12 for Medigoron

sonoftunk committed:
Check Tracker - Fixes #8 for Song from Saria

sonoftunk committed:
Check Tracker - Fixes #9 for Adult Shooting Gallery

sonoftunk committed:
Check Tracker - Fixes #10 for Composer Grave

sonoftunk committed:
Check Tracker - Fixes #14 for Song from Ocarina of Time

sonoftunk committed:
Check Tracker - Fixes #16 for Shooting the Sun

sonoftunk committed:
Check Tracker - Fixes #19 for Darunia's Joy

sonoftunk committed:
Save Editor - Adds missing label for Treasure Chest Game

sonoftunk committed:
Check Tracker - Fixes Great Faries checks

sonoftunk committed:
Check Tracker - Fixes Sheik at Temple

sonoftunk committed:
Check Tracker - Fixes Great Fairy auto-scroll to wrong location

sonoftunk committed:
Check Tracker - Updates blue warp checks
- Fixes Bongo Bongo and Twinrova checks

sonoftunk committed:
Check Tracker - Fixes Dampe's Gravedigging Tour

sonoftunk committed:
Check Tracker - Fixes Prescription to King Zora

sonoftunk committed:
Check Tracker - Fixes Bazaar showing under wrong area on tracker between adult/child

sonoftunk committed:
Check Tracker - Fixes scroll for Hyrule Market and Bazaar

sonoftunk committed:
Check Tracker - Fixes giving Adult Trade Items shown on tracker: Broken Sword, Saw, Cojiro, Eyeball Froge, and Eyedrops

sonoftunk committed:
Cleans up item_location

sonoftunk committed:
Check Tracker - Item List
- Fixes item list not updating when changing seeds/options
- Separates item exclusion list from item check list
- Clean up comments

sonoftunk committed:
Check Tracker - Adds Skip functionality to buttons

sonoftunk committed:
Check Tracker - Refactors adult trade items to RandomizerInf

sonoftunk committed:
Check Tracker - Fixes an issue where any time an item is removed the tracker would cause an application crash

sonoftunk committed:
Check Tracker - Item names are now dependent on file language

sonoftunk committed:
Check Tracker - Fixes Gerudo Card check not showing in Fast setting

sonoftunk committed:
Merges origin/develop into LocationTracker to resolve trivial conflict with libultraship submodule in randomizer_item_tracker.cpp

sonoftunk commented:
@PurpleHato I like the idea, but I'm a complete noob with ImGui and have no idea how I would get the Check Tracker to behave nicely with other tracker settings, like docking with other windows, the background/chroma colour, floating windows, only showing on button combo, and only enabled while paused. I tried, nothing behaved :) 

If desired, I can attempt implement as a simple window without any of these extra features. But if we do want the extra features, I feel it might make sense to leave it as an enhancement to the Item Tracker. 

Other than that one change, I feel comfortable to upgrade this from a Draft request. All remaining issues I know about I would consider low priority.

stratomaster64 commented:
> I like the idea, but I'm a complete noob with ImGui and have no idea how I would get the Check Tracker to behave nicely with other tracker settings, like docking with other windows, the background/chroma colour, floating windows, only showing on button combo, and only enabled while paused. I tried, nothing behaved :)

I'm surprised that copying from the item tracker doesn't quite work. I might consider making a PR on your branch to see about developing similar behavior with check tracker.

sonoftunk commented:
@stratomaster64 It worked a little TOO well. The features were completely synced up, even after splitting everything apart. So if I changed colour for one of them, both would change. If either were set to windowed mode, they would both show windowed, even if the other was set to floating. I'm not surprised considering my attempt was a literal copy-paste to a new file with all the unneeded bits removed. I tried tracking down what part was still the same, but after 2 hours I still got nowhere. I've loaded that code into a new branch over on my github https://github.com/HarbourMasters/Shipwright/commit/8c41cefb936fbd2641a87ae00c95c05df1f4ae56

sonoftunk committed:
Reorders headers to (hopefully) avoid CI build conflicts on linux/consoles

sonoftunk committed:
Explicitly defines vector header in randomizer.h to fix compile errors

sonoftunk committed:
Complete Refactor of Check Tracker ImGui

sonoftunk commented:
Completely redid the ImGui implementation. New code will be a lot more sustainable for all future changes. 

![image](https://user-images.githubusercontent.com/59073987/203201186-2762b8ea-a8ff-476a-93b7-a699b56c1aaa.png)

briaguya0 commented:
@sonoftunk any chance you could make a quick PR with just the formatting changes to `3drando/item_location.cpp` and `3drando/item_location.hpp`? i'd get that merged in super quick and it'd make it a lot easier to see the actual diff in those files (even using "hide whitespace" on gh it's not the cleanest diff)

sonoftunk committed:
Adds missing check tracker files to CMakeLists

sonoftunk committed:
Reverts the Item_location whitespace cleanup from ad10807c

sonoftunk commented:
@briaguya-ai I'll have to reapply the formatting from scratch for the PR. In the mean-time, I've revert'ed the formatting change, which should help determine what changed that wasn't formatting related.

briaguya0 commented:
@sonoftunk you should be able to cherry-pick ad10807 onto a new branch for a pr

sonoftunk commented:
@briaguya-ai The cherry-pick was my first attempt. Unfortunately, there's all but a few of the spoiler check changes done to the file prior to ad10807c. That's why it was much easier to revert it for now and manually fix the few changes I made afterwards. Now that I have that, it does give me an easy full list of the checks I did change, which I can use to cherry-pick ad10807c, and undo all the other changes manually. In hindsight, I should have started with the format changes....

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_objects.cpp:
```diff
@@ -917,7 +941,58 @@ void RandomizerCheckObjects::UpdateImGuiVisibility() {
+
+    return
+        (rcObj.rcArea != RCAREA_INVALID) && // don't show Invalid locations
+        (rcObj.vOrMQ != RCVORMQ_MQ) && // don't show MQ checks until we support MQ
```
we do support mq! (probably just update the comment right?)

sonoftunk replied:
MQ is not yet supported by the check tracker. I haven't had the chance to actually get the ROM loaded in to develop and test it. There's more than likely going to be some work to get the tracker to load different list of checks depending on which dungeons are MQ.

sonoftunk replied:
MQ checks are now shown on the tracker, depending on which dungeons are MQ. As per ef587be47aedd9f095f0210ff1aac9890c56ad7d

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_objects.cpp:
```diff
@@ -917,7 +941,58 @@ void RandomizerCheckObjects::UpdateImGuiVisibility() {
+        ((rcObj.rcType != RCTYPE_SMALL_KEY) || gRandomizeKeysanity != 1) && // 1 is the value for "vanilla" small keys
+        ((rcObj.rcType != RCTYPE_GF_KEY) || gRandomizeKeysanity != 0) && // 0 is the value for "vanilla" gf keys
+        ((rcObj.rcType != RCTYPE_BOSS_KEY) || gRandomizeBossKeysanity != 1) && // 1 is the value for "vanilla" boss keys
+        ((rcObj.rcType != RCTYPE_GANON_BOSS_KEY) || gRandomizeShuffleGanonBossKey != 0) && // 0 is the value for "vanilla" ganon's boss key
```
now that https://github.com/HarbourMasters/Shipwright/pull/1761 it'd be cool to replace these magic numbers with enum values

sonoftunk replied:
Resolved by fd88829aa6b2575460e6c11188892b3cce955f6e.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -0,0 +1,789 @@
+    // Logic for each check
+    for (auto& obj : checks)
+    {
+        // Second Column mode
+        //  if (twoColumnMode && areaChecksGotten[obj.rcArea] != areaChecksTotal[obj.rcArea]) {
+        //      //TODO draw left column
+        //      ImGui::TableNextColumn();
+        //      //TODO draw right column
+        //  }
+        //
```
is two column mode planned? if not then it'd be good to just remove these lines

sonoftunk replied:
Yes two column mode is still on my list to get done. In my opinion, not required for "Minimal Viable Product" of the check tracker.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -0,0 +1,789 @@
+void Teardown() {
+    initialized = false;
+    //delete checkStatus; //TODO may cause crashes due to race conditions during a draw cycle, but if you don't have it, then there's a memory leak when switching files
+    areasFullyChecked = 0;
+    checks.clear();
+    lastSaveCount = -1;
+    lastLocationChecked = RC_UNKNOWN_CHECK;
+    for (int i = 0; i < sizeof(areaChecksTotal); i++) {
+        areaChecksTotal[i] = 0;
+        areaChecksGotten[i] = 0;
+    }
+}
+
+bool ShouldUpdateChecks() {
+    // TODO eventually will need to be hooked into game elements rather than just save file
+    // TODO, saveCount isn't actually a thing. If not allowed to add it, will need to iterate over save bits and see if any changed 
+    // TODO, or enhance hooks, but that is a LUS change
+    //return lastSaveCount != gSaveContext.sohStats.saveCount;
+
+    return true;
+}
```
just to confirm, the current logic is to update the checks every frame?

sonoftunk replied:
Yes :( I couldn't find an obvious way to check to see if the game has been saved. If it's fine with you, I'll add the `saveCount` method. Very simple, low impact, good performance gain, and most importantly, I don't have to figure out how to do submodules :)

briaguya0 replied:
not everyone uses autosave, the check tracker shouldn't rely on the game being saved to update

sonoftunk replied:
Agreed! I've got a stub in place for Checked so that you eventually won't need to save to see it update, manual or auto regardless. This is also the reason ShouldUpdateChecks is a function, for all the possible hooks that could change the list of checked places; render functions, collected but unsaved checks (with or without reset), etc.

I've been going with the approach of "best is the enemy of the good" so far. I too want to see it come up by collection only, but that does increase the scope quite a lot, and might not be the task for a first time committer to SoH. I could try, but I make no promises I'm experienced enough to be up to the task :3

briaguya0 replied:
I'd prefer the check tracker get merged with a known performance issue (assuming it's only an issue when the tracker is open) than only function when someone has autosave enabled 

In that case, the performance improvements could be their own PR, and anyone who feels confident enough in their abilities and motivated enough to work on it could implement them

sonoftunk replied:
I've added an option for anyone experiencing performance issues with the tracker in commit 865d3cf4f72a06af122be87a1d9060514b90e01d. When enabled, it only confirms if one check has changed per draw cycle. This should act as a middle ground for any user who is experiencing performance problems and doesn't mind waiting a bit for the tracker to update after a save.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -0,0 +1,789 @@
+
+
+bool HasItemBeenCollected(RandomizerCheckObject obj) {
+    // TODO doesn't consider vanilla/MQ?
```
is this still an issue?

sonoftunk replied:
I haven't had the chance to actually get the ROM loaded in to develop and test it. It may work as is, but has not been confirmed.

sonoftunk replied:
MQ Support finally added as per ef587be47aedd9f095f0210ff1aac9890c56ad7d. However, It has not been confirmed that all checks are valid, but I'm quite hopeful it's all working as there weren't many checks that I needed to change for Vanilla dungeons, and the ones I did have to change were boss related. Will remove TODO once all checks are confirmed valid.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -0,0 +1,789 @@
+    if (lastLocationChecked == rcObj.rc)
+        txt = "* " + txt;
+    
+    //Second Column mode
+    //TODO
+    // if (twoColumnMode) ImGui::TableNextColumn();
```
same comment as before about 2 column, is it still planned?

sonoftunk replied:
Yes two column mode is still on my list to get done. In my opinion, not required for "Minimal Viable Product" of the check tracker.

briaguya0 reviewed (commented):
overall this is looking really cool! there are still quite a few TODOs in there (it's not quite clear if those are just leftover comments or not)

i left a couple comments with questions, really looking forward to this one!

sonoftunk commented:
#2004 submitted for the whitespace cleanup. Once merged into main branch, I have a commit ready to resolve merge conflicts. I'll hold off before committing in case it's not next in the list.

sonoftunk committed:
Check Tracker - Loading Fixes
- Fixes a bug where loading a save file a second time would add all the checks again
- Fixes a bug where the check tracker would not render at all until in a save file

sonoftunk committed:
Check Tracker - Fixes a bug where checks would sometimes fail to load

sonoftunk committed:
Check Tracker - Fixes duplicated checks when using Only on Pause or Only on Combo

sonoftunk committed:
Check Tracker - Works around issue where single characters don't show in EnhancementCombobox, and fixes a label for combo buttons held option

sonoftunk committed:
Merge origin/develop into LocationTracker for conflict resolution between stat tracker and upstream formatting changes to item locations

sonoftunk committed:
Check Tracker - Applies all Check Tracker changes to Item Location

sonoftunk committed:
Check Tracker - Updated Check List
- Adds Link's pocket, either in KF or HM depending on starting age
- Supports Scrubs and Merchants
- Uses enums instead of magic numbers
- Updates DC Scrub Short Names for MQ

sonoftunk committed:
Check Tracker - Adds a performance mode option

sonoftunk committed:
Check Tracker - List Fixes (requires new rando generation)
- Fixes Frogs Rupees always visible
- Fixes Kokiri Sword Chest not showing
- Fixes Weird Egg not showing

sonoftunk committed:
Check Tracker - MQ List basic support

sonoftunk committed:
Check Tracker - MQ/Vanilla anti-spoilers

sonoftunk committed:
Check Tracker - Collecting a compass now spoils the area

sonoftunk committed:
Check Tracker - Basic Vanilla Support and fixes
- Vanilla items show in check tracker, even if many won't auto check
- Fixes an issue with Gerudo Fortress Keys
- Refactors IsVisibleInCheckTracker to randomizer_check_tracjer

sonoftunk committed:
Merges origin/develop into LocationTracker to resolve conflict with randomizer_check_objects.cpp

sonoftunk committed:
Check Tracker - Fixes merge conflict from 08a487f35e1d760ab9e69b932edbe4da6bcd9edb

sonoftunk committed:
Check Tracker - Fixes tracker loading as Floating when default value is set to Windowed

sonoftunk committed:
Merge origin/develop into LocationTracker to fix merge conflict with parsing randomizer file

sonoftunk committed:
Randomizer Location Exclusion List
- Adds Scrubs, Merchants, Shops. Adult Trade items, MQ Dungeons, and Bean Salesman
- Fixes Kokiri Sword Chest, Weird Egg, and Frog Song Rupees from misbehaving
- Updates Gerudo Fortress logic to be dependent on Open/Fast/Normal Fortress
- Removes Invalid Check
- Adds RO Enum for MQ settings
- Updates magic values to RO enum

sonoftunk committed:
Location Exclusion List - Scrubs that are always randomized now always show on the list

sonoftunk committed:
For now, removes some extra code only needed for a Check Tracker.

sonoftunk committed:
Randomizer Location Exclusion List - Fixes Eyeball Frog from showing up when Adult Trade Items was off

sonoftunk committed:
Merges origin/develop into Randomizer Location Exclusion List - Fixes Merge conflict caused by this PR and another PR both adding MQ dungeon count enums.

sonoftunk committed:
Merges origin/develop into Location Tracker to correct merge conflicts from Enum changes

sonoftunk committed:
Location Tracker - Updates Eyedrops to use RandomizerInf instead of Chest

sonoftunk committed:
Check Tracker - Fixes a regression from f9362994 which broke showing checks based on settings.

sonoftunk committed:
Check Tracker - Fixes a bug where Link's Pocket item wasn't acting as checked for area item count

sonoftunk committed:
Check Tracker - Fixes a bug where Locations would incorrectly say Vanilla due to the first check in the area being a non-MQ check

briaguya0 commented:
@sonoftunk looks like quite a bit of work has been done on this since I last reviewed it, is it ready for another review or are you still adding/testing stuff?

sonoftunk commented:
@briaguya-ai Mostly bug fixes and the features we talked about. I believe the only feature in those commits that we didn't discuss in the earlier comments was f9362994b93ce8a582519796690b1f103a6d0372 for some basic vanilla support.

I'll do a cleanup commit for all the remaining TODOs and comments that are no longer valid. I'll ping you when I'm ready for final review, which hopefully I should have time to finish up in the next day or so.

sonoftunk committed:
Check Tracker - Fixes a crash when Check Tracker is docked but not visible

briaguya0 commented:
@sonoftunk anything i can do to help get this across the finish line?

sonoftunk committed:
Check Tracker - Fixes areas being spoiled when loading different save files

sonoftunk committed:
Check Tracker - Only reset variables if they are initialized

sonoftunk committed:
Check Tracker - Cleans up old comments, unused variables, methods, and macros, adds language TODOs,

sonoftunk committed:
Merges origin/develop into Check Tracker to fix merge conflict with CosmeticsEditor.cpp (Breaking Change for rainbow in UI)

sonoftunk committed:
Check Tracker - Readds rainbow UI elements after Cosmetics Editor refactor

sonoftunk committed:
Check Tracker - Removal of flag lookup that was refactored, no longer needed header, and some trivial whitespace fixes.

sonoftunk commented:
@briaguya-ai Thanks for the offer! I've just finished up my last pass of code cleanup and bugfixing, and it is ready for review again.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_objects.cpp:
```diff
@@ -894,11 +894,34 @@ std::map<RandomizerCheck, RandomizerCheckObject> RandomizerCheckObjects::GetAllR
+            rcAreaBySceneID[rcObject.sceneId] = rcObject.rcArea;
+        }
+        //Add checkless Hyrule Market areas to the area return
+        for (int id = (int)SCENE_ENTRA; id <= (int)SCENE_MARKET_RUINS; id++)
+            rcAreaBySceneID[(SceneID)id] = RCAREA_MARKET;
```
nitpick:
```suggestion
        for (int id = (int)SCENE_ENTRA; id <= (int)SCENE_MARKET_RUINS; id++) {
            rcAreaBySceneID[(SceneID)id] = RCAREA_MARKET;
        }
```

or 

```suggestion
        for (int id = (int)SCENE_ENTRA; id <= (int)SCENE_MARKET_RUINS; id++)  rcAreaBySceneID[(SceneID)id] = RCAREA_MARKET;
```

sonoftunk replied:
Went with option 1 with commit 50ac993d08a81e0e4405a00873fd84cf45713294

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_objects.cpp:
```diff
@@ -894,11 +894,34 @@ std::map<RandomizerCheck, RandomizerCheckObject> RandomizerCheckObjects::GetAllR
         locationIt.visibleInImgui = (
             (locationIt.vOrMQ != RCVORMQ_MQ) && // don't show MQ checks until we support MQ
-            (locationIt.rcType != RCTYPE_SHOP) && // don't show shop items until we have shopsanity
+            ((locationIt.rcType != RCTYPE_SHOP) || CVar_GetS32("gRandomizeShopsanity", 0) > 1) && // 1 is the value for "0 random items" for shop
```
```suggestion
            ((locationIt.rcType != RCTYPE_SHOP) || (
                (CVar_GetS32("gRandomizeShopsanity", RO_SHOPSANITY_OFF) != RO_SHOPSANITY_OFF) &&
                (CVar_GetS32("gRandomizeShopsanity", RO_SHOPSANITY_OFF) != RO_SHOPSANITY_ZERO_ITEMS)
            )) &&
```

sonoftunk replied:
In reality, I probably should be removing code changes to this method entirely from this code, and only apply them via #2070.

sonoftunk replied:
Resolved with merge (with expected conflicts from #2070) ee09aa07ee81ad2c1fe48c752c3c971442272f8d

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_objects.cpp:
```diff
@@ -917,7 +940,12 @@ void RandomizerCheckObjects::UpdateImGuiVisibility() {
-            ((locationIt.rcType != RCTYPE_GANON_BOSS_KEY) || CVar_GetS32("gRandomizeShuffleGanonBossKey", RO_GANON_BOSS_KEY_VANILLA) != RO_GANON_BOSS_KEY_VANILLA)
+            ((locationIt.rcType != RCTYPE_GANON_BOSS_KEY) || CVar_GetS32("gRandomizeShuffleGanonBossKey", RO_GANON_BOSS_KEY_VANILLA) != RO_GANON_BOSS_KEY_VANILLA) &&
+            ((locationIt.rcType != RCTYPE_GF_KEY && locationIt.rc != RC_GF_GERUDO_MEMBERSHIP_CARD) ||
+                (CVar_GetS32("gRandomizeGerudoFortress", 0) == 2 && locationIt.rcType != RCTYPE_GF_KEY && locationIt.rc != RC_GF_GERUDO_MEMBERSHIP_CARD) || //2 is the value for "open" gerudo's fortress
+                (CVar_GetS32("gRandomizeGerudoFortress", 0) == 1 && (locationIt.rc == RC_GF_NORTH_F1_CARPENTER || locationIt.rc == RC_GF_GERUDO_MEMBERSHIP_CARD)) || //1 is the value for "fast" gerudo's fortress
+                (CVar_GetS32("gRandomizeGerudoFortress", 0) == 0) //0 is the value for "normal" gerudo's fortress
```
```suggestion
                (CVar_GetS32("gRandomizeGerudoFortress", RO_GF_NORMAL) == RO_GF_OPEN && locationIt.rcType != RCTYPE_GF_KEY && locationIt.rc != RC_GF_GERUDO_MEMBERSHIP_CARD) ||
                (CVar_GetS32("gRandomizeGerudoFortress", RO_GF_NORMAL) == RO_GF_FAST && (locationIt.rc == RC_GF_NORTH_F1_CARPENTER || locationIt.rc == RC_GF_GERUDO_MEMBERSHIP_CARD)) ||
                (CVar_GetS32("gRandomizeGerudoFortress", RO_GF_NORMAL) == RO_GF_NORMAL)
```

sonoftunk replied:
In reality, I probably should be removing code changes to this method entirely from this code, and only apply them via #2070.

briaguya0 replied:
i just merged #2070 so hopefully when you pull latest dev in this stuff in this diff will clean itself up

sonoftunk replied:
Resolved with merge (with expected conflicts from #2070) ee09aa07ee81ad2c1fe48c752c3c971442272f8d

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -0,0 +1,1046 @@
+
+                if (isThisAreaSpoiled) {
+                    if (showVOrMQ && RandomizerCheckObjects::AreaIsDungeon(obj.rcArea)) {
+                        if (OTRGlobals::Instance->gRandomizer->masterQuestDungeons.contains(dungeonSceneLookupByArea[obj.rcArea]))
```
since `dungeonSceneLookupByArea` is only being used here, it might be nice to refactor that into a method with a switch/case or something, it's pretty hard to read that array as-is with the first 20 entries just being `SCENE_ID_MAX`

sonoftunk replied:
Refactored in commit 35b0a6033ab28067afd707ed2cb29a01e17d723c

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -0,0 +1,1046 @@
+    ImGui::End();
+}
+
+uint8_t gRandomizeShopsanity;
+uint8_t gRandomizeShuffleTokens;
+uint8_t gRandomizeShuffleBeans;
+uint8_t gRandomizeShuffleScrubs;
+uint8_t gRandomizeShuffleMerchants;
+uint8_t gRandomizeShuffleCows;
+uint8_t gRandomizeShuffleAdultTrade;
+uint8_t gRandomizeShuffleKokiriSword;
+uint8_t gRandomizeShuffleWeirdEgg;
+uint8_t gRandomizeShuffleGerudoCard;
+uint8_t gRandomizeShuffleFrogSongRupees;
+uint8_t gRandomizeStartingMapsCompasses;
+uint8_t gRandomizeKeysanity;
+uint8_t gRandomizeGerudoFortressKeys;
+uint8_t gRandomizeBossKeysanity;
+uint8_t gRandomizeShuffleGanonBossKey;
+uint8_t gRandomizeGerudoFortress;
```
not the biggest fan of these being global, maybe make them static in `IsVisibleInCheckTracker` instead?

sonoftunk replied:
My native language is C#, still getting used to variable scope problems like this. In C#, these would inherently be private class variables and wouldn't be visible outside of this file. I will indeed fix up the variable scope.

briaguya0 replied:
this file doesn't have a class, so while these variables would only be available in the scope of this file (unless someone tried to `#include` the `.cpp` somewhere), it still feels odd having them here

sonoftunk replied:
Resolved by wrapping everything in a namespace with f34934531de18172afa3d0ba7c8db5275b048bb9.

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -0,0 +1,1046 @@
+uint8_t gRandomizeShuffleGanonBossKey;
+uint8_t gRandomizeGerudoFortress;
+
+void SetVanillaSettings() {
+    gRandomizeShopsanity = RO_SHOPSANITY_OFF;
+    gRandomizeShuffleTokens = RO_TOKENSANITY_ALL;
+    gRandomizeShuffleBeans = RO_GENERIC_YES;
+    gRandomizeShuffleScrubs = RO_GENERIC_YES;
+    gRandomizeShuffleMerchants = RO_GENERIC_YES;
+    gRandomizeShuffleCows = RO_GENERIC_NO;
+    gRandomizeShuffleAdultTrade = RO_GENERIC_YES;
+    gRandomizeShuffleKokiriSword = RO_GENERIC_YES;
+    gRandomizeShuffleWeirdEgg = RO_GENERIC_YES;
+    gRandomizeShuffleGerudoCard = RO_GENERIC_YES;
+    gRandomizeShuffleFrogSongRupees = RO_GENERIC_NO;
+    gRandomizeStartingMapsCompasses = RO_DUNGEON_ITEM_LOC_ANYWHERE;
+    gRandomizeKeysanity = RO_DUNGEON_ITEM_LOC_ANYWHERE;
+    gRandomizeGerudoFortressKeys = RO_GERUDO_KEYS_VANILLA;
+    gRandomizeBossKeysanity = RO_DUNGEON_ITEM_LOC_ANYWHERE;
+    gRandomizeShuffleGanonBossKey = RO_GANON_BOSS_KEY_ANYWHERE;
+    gRandomizeGerudoFortress = RO_GF_NORMAL;
+}
```
this just feels strange (in vanilla beans aren't shuffled), i know the logic of `IsVisibleInCheckTracker` relies on that, but i think it'd actually be cleaner to just have 2 blocks in there depending on if rando is enabled or not

sonoftunk replied:
Cleaned up this with 69719abf7054f0d0fba3a3b8c55f94f414c9db7e

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -0,0 +1,1046 @@
+    gRandomizeGerudoFortress = RO_GF_NORMAL;
+}
+
+void SetRandomizerSettings() {
+    gRandomizeShopsanity =              OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_SHOPSANITY);
+    gRandomizeShuffleTokens =           OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_SHUFFLE_TOKENS);
+    gRandomizeShuffleBeans =            OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_SHUFFLE_MAGIC_BEANS);
+    gRandomizeShuffleScrubs =           OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_SHUFFLE_SCRUBS);
+    gRandomizeShuffleMerchants =        OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_SHUFFLE_MERCHANTS);
+    gRandomizeShuffleCows =             OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_SHUFFLE_COWS);
+    gRandomizeShuffleAdultTrade =       OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_SHUFFLE_ADULT_TRADE);
+    gRandomizeShuffleKokiriSword =      OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_SHUFFLE_KOKIRI_SWORD);
+    gRandomizeShuffleWeirdEgg =         OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_SHUFFLE_WEIRD_EGG);
+    gRandomizeShuffleGerudoCard =       OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_SHUFFLE_GERUDO_MEMBERSHIP_CARD);
+    gRandomizeShuffleFrogSongRupees =   OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_SHUFFLE_FROG_SONG_RUPEES);
+    gRandomizeStartingMapsCompasses =   OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_STARTING_MAPS_COMPASSES);
+    gRandomizeKeysanity =               OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_KEYSANITY);
+    gRandomizeGerudoFortressKeys =      OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_GERUDO_KEYS);
+    gRandomizeBossKeysanity =           OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_BOSS_KEYSANITY);
+    gRandomizeShuffleGanonBossKey =     OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_GANONS_BOSS_KEY);
+    gRandomizeGerudoFortress =          OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_GERUDO_FORTRESS);
+}
+
+bool IsVisibleInCheckTracker(RandomizerCheckObject rcObj) {
+    return
```
instead of having this as a method, you could just have a static `isInitialized` bool in `IsVisibleInCheckTracker` and then move the reading RSK logic into an `if` in there

if you want to get really fancy with it, you could simplify the checks in `IsVisibleInCheckTracker` a bit. for example, instead of having
```cpp
gRandomizeShopsanity = RO_SHOPSANITY_OFF;
```
```cpp
gRandomizeShopsanity =              OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_SHOPSANITY);
```
```cpp
(rcObj.rcType != RCTYPE_SHOP || gRandomizeShopsanity > RO_SHOPSANITY_ZERO_ITEMS) &&
```
you could have
```cpp
static bool shouldShowShopsanity = gSaveContext.n64ddFlag ? (
    OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_SHOPSANITY) != RO_SHOPSANITY_OFF &&
    OTRGlobals::Instance->gRandomizer->GetRandoSettingValue(RSK_SHOPSANITY) != RO_SHOPSANITY_ZERO_ITEMS) : false;
```


```c
rcObj.rcType != RCTYPE_SHOP || shouldShowShopsanity
```

sonoftunk replied:
The variable name structure was a holdover from when it was originally using the CVars to get the settings (which isn't true, it has to use the settings from the save file), and it was just a direct match to the CVar string to simplify remembering. And the layout of code was a pre-optimization to avoid having to always get the rando settings. 

IMO, having it separate makes it more readable, but I can try to refactor to 1 or 2 methods that handle all 4 blocks of code, and throw some comments for the readability.

briaguya0 replied:
if you want to just have `IsVisibleInCheckTracker` that calls `IsVisibleInVanillaCheckTracker` or `IsVisibleInRandoCheckTracker` or something that could work, my main concern from a readability stand point was what felt like hacky use of randomizer settings values to achieve desired behavior in vanilla

sonoftunk replied:
Refactored for readability with 69719abf7054f0d0fba3a3b8c55f94f414c9db7e

briaguya0 replied:
couple of things i noticed in the logic (out of scope for this PR)
* skulltulas will appear on the tracker when
  * playing vanilla
  * rando'd with tokens shuffled
this means the tracker won't work for going through to find any missing tokens when not playing tokensanity

fortress key checks won't appear in vanilla, because `fortressNormal` is set to false

this: `//Bug: the above will spoil that everything is vanilla if the random count rolled 0.`

 it'd probably be good to make an issue documenting this stuff so it can be addressed in a follow up PR at some point

sonoftunk replied:
@briaguya-ai Skulltullas appearing in vanilla is actually intentional, as my goal for vanilla thus far has been to show all checks that are required for the completionists out there who want to 100% the game.

briaguya0 replied:
@sonoftunk 
> Skulltullas appearing in vanilla is actually intentional

yes, that makes sense, the part that doesn't make sense is the lack of ability to track skull token locations in rando when not playing with tokensanity (this is handled in other trackers by having a toggle for show/hide skulltulas)

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -0,0 +1,1046 @@
+    if (gPlayState == nullptr || gSaveContext.fileNum < 0 || gSaveContext.fileNum > 2)
+        return;
+
+    if (gSaveContext.n64ddFlag)
+        SetRandomizerSettings();
+    else
+        SetVanillaSettings();
```
with my earlier suggestions this could be removed

sonoftunk replied:
Removed with refactor from 69719abf7054f0d0fba3a3b8c55f94f414c9db7e

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -0,0 +1,1046 @@
+    else if (checkStatusMap.find(i.rc)->second > checkStatusMap.find(j.rc)->second)
+        return false;
+
+    if (i.rcShortName < j.rcShortName)
+        return true;
+    if (i.rcShortName > j.rcShortName)
+        return false;
```
is there a way to avoid string comparisons here?

sonoftunk replied:
The goal of the string comparisons is to show randomizer checks in a deterministic sort order. I can replace the sorting with the RandomizerCheck, which will still give determinism.

sonoftunk replied:
Resolved by 12749935f6260b5e6cb057826e9caa12653d5689

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -0,0 +1,1046 @@
+        case SpoilerCollectionCheckType::SPOILER_CHK_GOLD_SKULLTULA:
+            return GET_GS_FLAGS(scene) & flag;
+        case SpoilerCollectionCheckType::SPOILER_CHK_INF_TABLE:
+            // Magic to flip an index `flag` to a lookup for 16bit big endian integers. Probably an easier way.....
+            shift = 7 - (flag % 8) + ((flag % 16) / 8) * 8;
```
since this is used in 2 places, it'd be nice to turn this into a macro

then if someone comes up with an easier/better way it only needs to be changed in one place

sonoftunk replied:
Refactored as a part of fcdf83bc2c773b5b55ae2bc79b70a2b161a238ed

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer_check_tracker.cpp:
```diff
@@ -0,0 +1,1046 @@
+void ImGuiDrawTwoColorPickerSection(const char* text, const char* cvarMainName, const char* cvarExtraName,
+                                 Color_RGBA8& main_color, Color_RGBA8& extra_color, Color_RGBA8& main_default_color,
+                                 Color_RGBA8& extra_default_color, const char* cvarHideName) {
+    Color_RGBA8 cvarMainColor = CVar_GetRGBA(cvarMainName, main_default_color);
+    Color_RGBA8 cvarExtraColor = CVar_GetRGBA(cvarExtraName, extra_default_color);
+    main_color = cvarMainColor;
+    extra_color = cvarExtraColor;
+
+    if (ImGui::CollapsingHeader(text)) {
+        if (*cvarHideName != '\0') {
+            std::string label = cvarHideName;
+            label += "##Hidden";
+            ImGui::PushID(label.c_str());
+            UIWidgets::EnhancementCheckbox("Hidden", cvarHideName, false,
+                                           "When active, checks will hide by default when updated to this state. Can "
+                                           "be overriden with the \"Show Hidden Items\" option.");
+            ImGui::PopID();
+        }
+        if (ImGui::BeginTable(text, 2, ImGuiTableFlags_BordersH | ImGuiTableFlags_BordersV | ImGuiTableFlags_Hideable)) {
+            ImGui::TableNextRow();
+            ImGui::TableNextColumn();
+            ImGui::PushItemWidth(ImGui::GetContentRegionAvail().x);
+            if (UIWidgets::EnhancementColor("Check", cvarMainName, Color_RGBA8_to_ImVec4(main_color), Color_RGBA8_to_ImVec4(main_default_color))) {
+                main_color = CVar_GetRGBA(cvarMainName, main_default_color);
+            };
+            ImGui::PopItemWidth();
+            
+            ImGui::TableNextColumn();
+            ImGui::AlignTextToFramePadding();
+            ImGui::PushItemWidth(ImGui::GetContentRegionAvail().x);
+            if (UIWidgets::EnhancementColor("Details", cvarExtraName, Color_RGBA8_to_ImVec4(extra_color),
+                                            Color_RGBA8_to_ImVec4(extra_default_color))) {
+                main_color = CVar_GetRGBA(cvarExtraName, extra_default_color);
+            }
+            ImGui::PopItemWidth();
+
+            ImGui::EndTable();
+        }
+    }
+}
```
i feel like these exist in other places already, could this be done without duplicated code?

sonoftunk replied:
The rainbow code is a modification of Cosmetic Editor's code for rainbows. With a recent refactor of it from develop, I wasn't able to use Cosmetic Editor directly any more. I can look at refactoring both Cosmetic Editor and Check Tracker to both use a new standard rainbow generator? Or take a second crack at jamming in my UI code into the Cosmetics?

The colour CVar save/load can also be refactored to just use the CVar_GetRGBA directly instead of the floats. It was only there to avoid having to rewrite the background colour/opacity. 

I couldn't find a standard implementation of the colour conversion method between RGBA8 and ImVec4 that the ImGui Widgets use. I can directly inline the code if the method is a problem.

Similarly, I wasn't able to see how I could use Cosmetic Editor's grid drawing to my advantage here in the check tracker options. I'll take another look for getting it to work and/or look to see if there's any other places

briaguya0 replied:
i figure worst case scenario just adding a todo and making an issue to remove the copypasta could work, but hopefully you can figure something out!

sonoftunk replied:
Commit e9e0a36add27b24573f62fd1bf7483caee3c7318 fixes a bug in UIWidgets where alpha-enabled EnhancementColor was practically useless, and 47af3dfd35a4e72310de48a5fbe8b79a12214542 refactors to get rid of the inline methods. Although, it doesn't seem much better. I would argue that Color_RGBA8_to_ImVec4 kept the code cleaner, and I couldn't find an alternative anywhere. 

As for the other points on the Cosmetic Editor, I still don't see a way of easily getting the two code-bases to cooperate. The two colour picker cannot be replaced by DrawCosmeticGroup, and the rainbow with CosmeticsUpdateTick, without major rework two both modules.

briaguya0 replied:
let's make a follow up issue for it, combining the codebases feels out of scope for this

as for https://github.com/HarbourMasters/Shipwright/commit/e9e0a36add27b24573f62fd1bf7483caee3c7318, i'm guessing you mean it wasn't being used before and the incorrect logic wasn't noticed?

briaguya0 reviewed (changes requested):
overall this is looking pretty good! there are a few places where i left suggestions that could clean up the logic a little, and a few places where i have some questions


looking forward to the next pass!

sonoftunk committed:
Check Tracker - Refactors Index to Bitmask code to a macro

sonoftunk committed:
Check Tracker - Changes ordering from alphabetical to RandomizerCheck ordering

sonoftunk committed:
Check Tracker - Wraps everything in a namespace to avoid global variable usage

sonoftunk committed:
Check Tracker - Explicitly defines for-loop scope

sonoftunk committed:
Check Tracker - refactors dungeon lookup from static array to function

sonoftunk committed:
Fixes a bug when using EnhancementColor with alpha

sonoftunk committed:
Check Tracker - Refactors all colour to use direct conversions

sonoftunk committed:
Check Tracker - refactors settings loading and check visibility for readability

sonoftunk committed:
Merge origin/develop into Check Tracker for conflicts related to changes made in #2070

sonoftunk commented:
@briaguya-ai Thanks for the feedback cycle! I went ahead and applied code changes for the majority of your feedback. I left some conversations unresolved as I believe they will definitely need a second look to be sure they are up to snuff.

sonoftunk committed:
Check Tracker - Fixes a bug where the wrong label showed on the Check Tracker BG Color

sonoftunk committed:
UIWidgets - Adds Alpha bar to EnhancementColor when using alpha parameter

briaguya0 reviewed (approved):
Good stuff!

I left a couple final comments, but nothing that needs to be addressed in this PR.

Let me know when you're ready and we'll :shipit:!

briaguya0 commented:
improvements to be implemented later: https://github.com/HarbourMasters/Shipwright/issues/2131
